### PR TITLE
Make HTML template functions support context

### PR DESCRIPTION
**Pull request by wxiaoguang**
# Background

Golang template is not friendly for large projects, and Golang template team is quite slow, related: 
* `https://github.com/golang/go/issues/54450`

Without upstream support, we can also have our solution to make HTML template functions support context.

It helps a lot, the above Golang template issue `#54450` explains a lot:

1. It makes `{{Locale.Tr}}` could be used in any template, without passing unclear `(dict "root" . )` anymore.
2. More and more functions need `context`, like `avatar`, etc, we do not need to do `(dict "Context" $.Context)` anymore.
3. Many request-related functions could be shared by parent&children templates, like "user setting" / "system setting"

See the test `TestScopedTemplateSetFuncMap`, one template set, two `Execute` calls with different `CtxFunc`.

# The Solution

Instead of waiting for upstream, this PR re-uses the escaped HTML template trees, use `AddParseTree` to add related templates/trees to a new template instance, then the new template instance can have its own FuncMap , the function calls in the template trees will always use the new template's FuncMap.

`template.New` / `template.AddParseTree` / `adding-FuncMap` are all quite fast, so the performance is not affected.

The details:

1. Make a new `html/template/Template` for `all` templates
2. Add template code to the `all` template
3. Freeze the `all` template, reset its exec func map, it shouldn't execute any template.
4. When a router wants to render a template by its `name`
    1. Find the `name` in `all`
    2. Find all its related sub templates
    3. Escape all related templates (just like what the html template package does)
    4. Add the escaped parse-trees of related templates into a new (scoped) `text/template/Template`
    5. Add context-related func map into the new (scoped) text template
    6. Execute the new (scoped) text template
    7. To improve performance, the escaped templates are cached to `template sets`

# FAQ

## There is a `unsafe` call, is this PR unsafe?

This PR is safe. Golang has strict language definition, it's safe to do so: https://pkg.go.dev/unsafe#Pointer (1) Conversion of a *T1 to Pointer to *T2

## What if Golang template supports such feature in the future?

The public structs/interfaces/functions introduced by this PR is quite simple, the code of `HTMLRender` is not changed too much. It's very easy to switch to the official mechanism if there would be one.

## Does this PR change the template execution behavior?

No, see the tests (welcome to design more tests if it's necessary)

### CODE REVIEW

**Review by silverwind — APPROVED** [review]

File: modules/templates/scopedtmpl/scopedtmpl.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+	return (*T)(unsafe.Pointer(ptr))
+}
+
+func newScopedTemplateSet(all *template.Template, name string) (*scopedTemplateSet, error) {
+	targetTmpl := all.Lookup(name)
+	if targetTmpl == nil {
+		return nil, fmt.Errorf("template %q not found", name)
+	}
+	if err := escapeTemplate(targetTmpl); err != nil {
+		return nil, fmt.Errorf("template %q has an error when escaping: %v", name, err)
```
Comment (wolfogre) [inline modules/templates/scopedtmpl/scopedtmpl.go]:
```suggestion
		return nil, fmt.Errorf("template %q has an error when escaping: %w", name, err)
```

File: modules/templates/scopedtmpl/scopedtmpl.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+					subTmpl := all.Lookup(subName)
+					if subTmpl == nil {
+						// HTML template will add some internal templates like "$delimDoubleQuote" into the text template
+						ts.textTemplates[subName] = textTmpl.Lookup(subName)
+					} else if subTmpl.Tree == nil || subTmpl.Tree.Root == nil {
+						collectErr = fmt.Errorf("template %q has no tree, it's usually caused by broken templates", subName)
+					} else {
+						ts.htmlTemplates[subName] = subTmpl
+						if err := escapeTemplate(subTmpl); err != nil {
+							collectErr = fmt.Errorf("template %q has an error when escaping: %v", subName, err)
```
Comment (wolfogre) [inline modules/templates/scopedtmpl/scopedtmpl.go]:
```suggestion
							collectErr = fmt.Errorf("template %q has an error when escaping: %w", subName, err)
```

Comment (wxiaoguang) [reply modules/templates/scopedtmpl/scopedtmpl.go]: Oops, that's written by my Copilot, not by me.  😂

(Yes, just kidding, indeed I didn't find the problem, so still my mistake 😁 )

**Review by wolfogre — APPROVED** [review]

**Comment by silverwind** [comment]
.

**Pull request merged**